text/3892-complex-numbers.md

```diff
@@ -0,0 +1,227 @@
+- Non-generic primitive types: These are, obviously, the most obvious and practical solution. However, if we implemented lots of such types, then we would not be able to expand for `f16` and `f128` support without repeating the code already implemented. It would be extremely repetitive and tedious to document new types and their behavior, even if we used macros to generate implementations
+
+## Prior art
+[prior-art]: #prior-art
+
+FORTRAN, C, C++, Go, Perl and Python all have complex types implemented in the standard library or as a primitive. This clearly appears to be an important feature many languages have.
+For example, in Python:
+```py
+complex_num = 1 + 2j
+complex_second = 3 + 4j
+print(complex_num * complex_second)
+```
+or in C:
+```c
+float _Complex cmplx = 1 + 2*I;
+float _Complex two_cmplx = 3 + 4*I;
+printf("%.1f%+.1fi\n", creal(cmplx * two_cmplx), cimag(cmplx * two_cmplx));
+```
+Even in Rust, it has been discussed two times in IRLO:
+- [First discussion](https://internals.rust-lang.org/t/c-compatible-complex-types-using-traits/13757)
+- [Second discussion](https://internals.rust-lang.org/t/standard-complex-number-in-std-library/23748)
+
+Many crates, like `num-complex` also provide this feature, though it is not FFI-safe.
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Should this type be in `core::ffi`? This type's purpose is mostly FFI, but it might be useful in library contexts as well, so I am not sure if we should place it in `core::ffi`.
+
+## Future possibilities
+[future-possibilities]: #future-possibilities
+
+- Maybe later on, we can think of adding a special custom suffix for complex numbers (`1+2j` for example), and using that as a simpler way of writing complex numbers if this RFC is accepted? This is very similar to how most languages implement complex numbers? Or perhaps we could consider a constant:
+```rust
+impl<T: Float> Complex<T: Float> {
+  const I: T = Complex::new(T::zero(), T::one());
+}
+```
+where `zero` and `one` is implemented on the `Float` trait similar to `num_traits`?
+Or maybe we could have a method on normal numbers:
+```rust
+// for example
+impl f32 {
+  fn i(self) -> Complex<f32> {
+    Complex::new(0, self)
+  }
+}
+```
+that could help simplify the life of people who otherwise would have to keep writing `Complex::new()`?
+- Should we support Imaginary eventually? This RFC doesn't cover it, but I think we can do this later in another RFC.
+- Eventually we may support Gaussian integers (an extension of the real integers) which have a Euclidean division procedure with remainder. GCC has these, and we could theoretically eventually support these integers alongside GCC FFI.
```

Comment:
FWIW, a recent issue raised that `num-complex`'s remainder is not ideal for their purpose: https://github.com/rust-num/num/issues/445

Although as I commented there, it can be seen similarly to the `Rem` vs. `rem_euclid` methods for integers in the standard library.